okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java

```diff
@@ -362,37 +365,64 @@ public String decodeQuery() {
       String name = queryNamesAndValues.get(i);
       String value = queryNamesAndValues.get(i + 1);
       if (i > 0) result.writeByte('&');
-      percentDecode(result, name, 0, name.length());
+      percentDecode(result, name, 0, name.length(), true);
       if (value != null) {
         result.writeByte('=');
-        percentDecode(result, value, 0, value.length());
+        percentDecode(result, value, 0, value.length(), true);
       }
     }
     return result.readUtf8();
   }
 
+  public int querySize() {
+    return queryNamesAndValues != null ? queryNamesAndValues.size() / 2 : 0;
+  }
+
   /**
    * Returns the first query parameter named {@code name} decoded using UTF-8, or null if there is
    * no such query parameter.
    */
   public String queryParameter(String name) {
-    throw new UnsupportedOperationException(); // TODO(jwilson).
+    if (queryNamesAndValues == null) return null;
+    String encodedName = canonicalize(name, QUERY_COMPONENT_ENCODE_SET, false, true);
+    for (int i = 0, size = queryNamesAndValues.size(); i < size; i += 2) {
+      if (encodedName.equals(queryNamesAndValues.get(i))) {
+        String value = queryNamesAndValues.get(i + 1);
+        return value != null ? percentDecode(value, true) : null;
+      }
+    }
+    return null;
```

Comment:
This is ambiguous between "not found" and "found but value was null".

When is the value `null`? `http://host/?foo`? `http://host/?foo=`? I would think the last one would be empty string. Not sure what the behavior of the former would be.